scalanet/discovery/src/io/iohk/scalanet/discovery/ethereum/v4/DiscoveryNetwork.scala

```diff
@@ -0,0 +1,376 @@
+      // Timeout for collecting multiple potential Neighbors responses.
+      kademliaTimeout: FiniteDuration = 7.seconds,
+      // Max number of neighbours to expect.
+      kademliaBucketSize: Int = 16
+  )(implicit codec: Codec[Payload], sigalg: SigAlg, clock: Clock[Task]): Task[DiscoveryNetwork[A]] = Task {
+    new DiscoveryNetwork[A] with LazyLogging {
+
+      import DiscoveryRPC.ENRSeq
+      import Payload._
+
+      private val expirationMillis = messageExpiration.toMillis
+
+      private val currentTimeMillis = clock.realTime(MILLISECONDS)
+
+      private val maxNeighborsPerPacket = getMaxNeighborsPerPacket
+
+      private val localNodeAddress = toNodeAddress(peerGroup.processAddress)
+
+      /** Start a fiber that accepts incoming channels and starts a dedicated fiber
+        * to handle every channel separtely, processing their messages one by one.
+        * This is fair: every remote connection can be throttled independently
+        * of each other, as well as based on operation type by the `handler` itself.
+        */
+      override def startHandling(handler: DiscoveryRPC[(PublicKey, A)]): Task[CancelableF[Task]] =
+        for {
+          cancelToken <- Deferred[Task, Unit]
+          _ <- peerGroup
+            .nextServerEvent()
+            .withCancelToken(cancelToken)
+            .toIterant
+            .mapEval {
+              case ChannelCreated(channel, release) =>
+                handleChannel(handler, channel, cancelToken)
+                  .guarantee(release)
+                  .onErrorRecover {
+                    case ex: TimeoutException =>
+                    case NonFatal(ex) =>
+                      logger.error(s"Error handling channel from ${channel.to}: $ex")
+                  }
+                  .startAndForget
+
+              case _ =>
+                Task.unit
+            }
+            .completedL
+            .startAndForget
+          cancelable <- CancelableF[Task](cancelToken.complete(()))
+        } yield cancelable
+
+      private def handleChannel(
+          handler: DiscoveryRPC[(PublicKey, A)],
+          channel: Channel[A, Packet],
+          cancelToken: Deferred[Task, Unit]
+      ): Task[Unit] = {
+        channel
+          .nextMessage()
+          .withCancelToken(cancelToken)
+          .timeout(messageExpiration) // Messages older than this would be ignored anyway.
+          .toIterant
+          .mapEval {
+            case MessageReceived(packet) =>
+              currentTimeMillis.flatMap { timestamp =>
+                Packet.unpack(packet) match {
+                  case Attempt.Successful((payload, remotePublicKey)) =>
+                    payload match {
+                      case _: Payload.Response =>
+                        // Not relevant on the server channel.
+                        Task.unit
+
+                      case p: Payload.HasExpiration[_] if p.expiration < timestamp - expirationMillis =>
```

Comment:
Indeed the spec says:
> The expiration field present in all packets is supposed to prevent packet replay. Since it is an absolute time stamp, the node's clock must be accurate to verify it correctly. Since the protocol's launch in 2016 we have received countless reports about connectivity issues related to the user's clock being wrong.

Some kind of check seems prudent and I guess you can always increase the expiration to ridiculous values like 24 hours - I doubt anyone can participate in blockchain if their computer's clock is off by that much. I took the default 1 minute from trinity. Funnily if we increase the expiration to higher the other side will accept it as well as long as its not in the past.

I saw some discussions about how bad it is to require a centralised authority like NPT in a decentralised application for correctness; it sounds like a good idea to warn about clock drift even if it's disabled. I think it should be in another component though, something that the node starts. It could be relevant to certain consensus algorithms as well.

Reply:
Ah, I forgot that the the expiration check means if the other side is using a 1 minute expiration and we're off by more than a minute we can still ignore their messages, so increasing our timeout value doesn't help. Maybe there should be a setting to allow a maximum clock drift.

Reply:
Added a `maxClockDrift` setting, that can basically disable checks.